tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/CloneRepos.cs

```diff
@@ -0,0 +1,155 @@
+                return false;
+            }
+        }
+
+        private async Task<bool> CheckoutRepo(string path, string hash)
+        {
+            return await RunExternalGitCommand($"checkout {hash}") &&
+                   await RunExternalGitCommand($"submodule update --init --recursive");
+        }
+
+        private async Task<bool> CloneRepo(string url, string path)
+        {
+            return await RunExternalGitCommand($"clone --recursive {url} {path}");
+        }
+
+        private async void EnsureGit()
+        {
+            try
+            {
+                var gitExists = await RunExternalGitCommand("--version");
+            }
+            catch (Exception e)
+            {
+                Log.LogError("git does not appear to be installed or working.  See error message for details.");
+                Log.LogErrorFromException(e, true, true, null);
+                throw;
+            }
+        }
+
+        private static IEnumerable<RepoSpec> ReadSpecFile(string filePath)
+        {
+            var doc = XDocument.Load(filePath);
+            foreach (var dep in doc.Element("Dependencies").Element("ToolsetDependencies").Elements("Dependency").Concat(doc.Element("Dependencies").Element("ProductDependencies").Elements("Dependency")))
+            {
+                if (!string.IsNullOrWhiteSpace(dep.Element("SkipClone")?.Value) && dep.Element("SkipClone").Value.ToLowerInvariant() == "true")
+                {
+                    continue;
+                }
+                if (!string.IsNullOrWhiteSpace(dep.Element("RepoName")?.Value))
+                {
+                    yield return new RepoSpec(dep.Element("Uri").Value, dep.Element("RepoName").Value, dep.Element("Sha").Value);
+                }
+                else
+                {
+                    yield return new RepoSpec(dep.Element("Uri").Value, dep.Element("Sha").Value);
+                }
+            }
+        }
+
+        private class RepoSpec
+        {
+            internal string Url { get; set; }
+            internal string Name { get; set; }
+            internal string Hash { get; set; }
+
+            public RepoSpec(string url, string hash) : this(url, url.Substring(url.LastIndexOf('/') + 1), hash)
+            {
+            }
+
+            public RepoSpec(string url, string name, string hash)
+            {
+                this.Url = url;
+                this.Name = name;
+                this.Hash = hash;
+            }
+
+            public override bool Equals(object obj)
+            {
+                var other = obj as RepoSpec;
+                if (other == null)
+                {
+                    return false;
+                }
+                return other.Url == this.Url && other.Name == this.Name;
+            }
+
+            public override int GetHashCode()
+            {
+                return this.Url.GetHashCode() ^ this.Name.GetHashCode();
```

Comment:
Are these really necessary?

If so, maybe use tuple hashcode and equality to avoid implementing your own? `^` is also not a great hash combiner.

(Also, looks out of date.)